### PR TITLE
Add lark sheets connector

### PR DESCRIPTION
**Pull request by 7c00**
This pull introduces Lark Sheets Connector, which is inspired by [Google Sheets Connector](https://github.com/prestodb/presto/pull/16721). Though there are tiny differences between them. 

Developers could run `LarkSheetsQueryRunners#main`, and then follow the _QueryModel_ section from the [doc](https://github.com/7c00/presto/blob/presto-lark/presto-docs/src/main/sphinx/connector/larksheets.rst) to try the connector.

Test plan

- Integation test
- Unit test

```
== RELEASE NOTES ==

Connectors
* Introduce Lark sheets connector.

```

### CODE REVIEW

**Comment by 7c00** [comment]
Hello, @ChunxuTang @beinan. Could you help review this PR?

**Comment by rohanpednekar** [comment]
Thanks, @7c00 for your contribution, we were planning to request @ChunxuTang and @beinan as they helped us earlier to review the google sheet connector.

File: presto-lark-sheets/src/main/java/com/facebook/presto/lark/sheets/SheetsConfig.java
Comment (ChunxuTang) [inline presto-lark-sheets/src/main/java/com/facebook/presto/lark/sheets/SheetsConfig.java:25]: Just wondering what's the difference between `FEISHU` and `LARK`? This is the lark connector, but the default domain is `FEISHU`?

File: presto-lark-sheets/src/test/resources/testing-app-secret-base64
Comment (ChunxuTang) [inline presto-lark-sheets/src/test/resources/testing-app-secret-base64:1]: Is there a TTL for this secret? And is there an owner email account? There's a potential risk here that when the secret expires, it may break presto CI workflows.

**Review by ChunxuTang — COMMENTED** [review]
Hi @7c00, thanks a lot for your work! This looks to be a very nice feature, and the Presto community could benefit.
Just left some suggestions/comments. For the first round of review, I mainly focused on the coding styles. Thanks again for your contributions!

File: presto-lark-sheets/src/main/java/com/facebook/presto/lark/sheets/SheetsMetadata.java
Comment (7c00) [reply presto-lark-sheets/src/main/java/com/facebook/presto/lark/sheets/SheetsMetadata.java:133]: The parameter `desiredColumns` is definied by the SPI method [ConnectorMetadata#getTableLayouts](https://github.com/prestodb/presto/blob/d37ac0735773af0cd17b26fff1e2c9a9575734d8/presto-spi/src/main/java/com/facebook/presto/spi/connector/ConnectorMetadata.java#L113). So we could not remove it.

File: presto-lark-sheets/src/main/java/com/facebook/presto/lark/sheets/SheetsMetadata.java
Comment (7c00) [reply presto-lark-sheets/src/main/java/com/facebook/presto/lark/sheets/SheetsMetadata.java]: Yes. I have replaced Streams.stream with StreamSupport.stream.

File: presto-lark-sheets/src/main/java/com/facebook/presto/lark/sheets/SheetsMetadata.java
Comment (7c00) [reply presto-lark-sheets/src/main/java/com/facebook/presto/lark/sheets/SheetsMetadata.java]: Added some comments to the code for details. 
>    // Use prefixTableName (other than SheetInfo#getTitle) to create SchemaTableName
>    // in order to make queries like `DESC "@sheetId"` or `DESC "#1"` work

File: presto-lark-sheets/src/main/java/com/facebook/presto/lark/sheets/SheetsRecordCursor.java
Comment (7c00) [reply presto-lark-sheets/src/main/java/com/facebook/presto/lark/sheets/SheetsRecordCursor.java]: Removed.

File: presto-lark-sheets/src/main/java/com/facebook/presto/lark/sheets/SheetsSplit.java
Comment (7c00) [reply presto-lark-sheets/src/main/java/com/facebook/presto/lark/sheets/SheetsSplit.java]: Fixed.

File: presto-lark-sheets/src/main/java/com/facebook/presto/lark/sheets/api/SheetsApiFactory.java
Comment (7c00) [reply presto-lark-sheets/src/main/java/com/facebook/presto/lark/sheets/api/SheetsApiFactory.java]: Fixed.

File: presto-lark-sheets/src/main/java/com/facebook/presto/lark/sheets/SheetsConfig.java
Comment (7c00) [reply presto-lark-sheets/src/main/java/com/facebook/presto/lark/sheets/SheetsConfig.java]: Lark and Feishu are the same product but serving different users. Lark serves global users, while Feishu serves Chinese users. They have different SDK domains.

I have changed the default value to LARK.

File: presto-lark-sheets/src/test/resources/testing-app-secret-base64
Comment (7c00) [reply presto-lark-sheets/src/test/resources/testing-app-secret-base64:1]: There is no TTL for the secret. The secret is not bound to an email account. 
I wonder if we should disable `TestSheetsIntegration` by default as `TestGoogleSheets`.

File: presto-lark-sheets/src/main/java/com/facebook/presto/lark/sheets/SheetsRecordCursor.java
Comment (7c00) [reply presto-lark-sheets/src/main/java/com/facebook/presto/lark/sheets/SheetsRecordCursor.java:211]: The static method is placed here since its logic is not general but very sepecial to SheetsRecordCursor.

File: presto-lark-sheets/src/main/java/com/facebook/presto/lark/sheets/api/SheetsApiFactory.java
Comment (7c00) [reply presto-lark-sheets/src/main/java/com/facebook/presto/lark/sheets/api/SheetsApiFactory.java:214]: Moved the utility function like `loadAppSecret` to SheetsUtil.

For `toLarkDomain`, it translates SheetsConfig.Domain to Lark SDK Domain. As we encapsulated all SDK invocations into `SheetsApiFactory`, it would be better to keep `toLarkDomain` remain in this class. What do you think?

**Comment by 7c00** [comment]
@ChunxuTang Thanks a lot for your comments. I haved update the code. Could you please review it again?

File: presto-lark-sheets/src/main/java/com/facebook/presto/lark/sheets/SheetsMetadata.java
Comment (ChunxuTang) [reply presto-lark-sheets/src/main/java/com/facebook/presto/lark/sheets/SheetsMetadata.java:133]: Thanks for the clarification!

File: presto-lark-sheets/src/main/java/com/facebook/presto/lark/sheets/SheetsMetadata.java
Comment (ChunxuTang) [inline presto-lark-sheets/src/main/java/com/facebook/presto/lark/sheets/SheetsMetadata.java:289]: What if the column name is non-English? Will it throw an exception or just skip the column silently?

File: presto-lark-sheets/src/main/java/com/facebook/presto/lark/sheets/api/SheetsApiFactory.java
Comment (ChunxuTang) [reply presto-lark-sheets/src/main/java/com/facebook/presto/lark/sheets/api/SheetsApiFactory.java:214]: Sure, thanks for the clarification.

File: presto-lark-sheets/src/main/java/com/facebook/presto/lark/sheets/api/SheetsApiFactory.java
Comment (ChunxuTang) [inline presto-lark-sheets/src/main/java/com/facebook/presto/lark/sheets/api/SheetsApiFactory.java:82]: Just wondering as the `CachingSheetsApi` is implemented in a separate file, why do you prefer to implement the `SimpleSheetsApi` in the factory class? Any specific reasons?

**Review by ChunxuTang — COMMENTED** [review]
Hi @7c00, thanks for your revision!
Left a few minor suggestions/comments.

File: presto-lark-sheets/src/test/resources/testing-app-secret-base64
Comment (ChunxuTang) [reply presto-lark-sheets/src/test/resources/testing-app-secret-base64:1]: Yeah, I think we may need to disable it.
Hi @beinan, could you help disable this integration test by default?

File: presto-docs/src/main/sphinx/connector/larksheets.rst
Comment (7c00) [reply presto-docs/src/main/sphinx/connector/larksheets.rst:31]: Yes, it is a real app id.

File: presto-docs/src/main/sphinx/connector/larksheets.rst
Comment (7c00) [reply presto-docs/src/main/sphinx/connector/larksheets.rst]: Fixed.

File: presto-docs/src/main/sphinx/connector/larksheets.rst
Comment (7c00) [reply presto-docs/src/main/sphinx/connector/larksheets.rst:64]: No, it will not expire.

File: presto-lark-sheets/src/main/java/com/facebook/presto/lark/sheets/SheetsMetadata.java
Comment (7c00) [reply presto-lark-sheets/src/main/java/com/facebook/presto/lark/sheets/SheetsMetadata.java:289]: Presto supports UTF-8 identifier names. So for a non-english string, it can be used as a column name in Lark connector, but should be quoted as `"XXX"`to access the column.

File: presto-lark-sheets/src/test/java/com/facebook/presto/lark/sheets/TestSheetsPlugin.java
Comment (7c00) [reply presto-lark-sheets/src/test/java/com/facebook/presto/lark/sheets/TestSheetsPlugin.java]: Removed as it's not used any more.

File: presto-lark-sheets/src/test/java/com/facebook/presto/lark/sheets/TestSheetsUtil.java
Comment (7c00) [reply presto-lark-sheets/src/test/java/com/facebook/presto/lark/sheets/TestSheetsUtil.java:1]: Sure. Added tests for new functions here.

File: presto-lark-sheets/src/main/java/com/facebook/presto/lark/sheets/api/SheetsApiFactory.java
Comment (7c00) [reply presto-lark-sheets/src/main/java/com/facebook/presto/lark/sheets/api/SheetsApiFactory.java]: Possible some of my perference during development I guess. Have rafactored SimpleSheetsApi to a top level class to make it matched the code style of CachingSheetsApi.

**Comment by 7c00** [comment]
@ChunxuTang Thanks for your comments. :heart: 
Have updated the code. Could you please take a review again? Thanks in advance.

**Comment by ChunxuTang** [comment]
> @ChunxuTang Thanks for your comments. ❤️ Have updated the code. Could you please take a review again? Thanks in advance.

Sure! I'll take another round of review very soon this week.

File: presto-docs/src/main/sphinx/connector/larksheets.rst
Comment (ChunxuTang) [reply presto-docs/src/main/sphinx/connector/larksheets.rst:31]: Just wondering as this is a real ID, is it OK to expose it in an example file? Maybe just use `exampleAppId`?
Your call.

File: presto-lark-sheets/src/main/java/com/facebook/presto/lark/sheets/SheetsMetadata.java
Comment (ChunxuTang) [reply presto-lark-sheets/src/main/java/com/facebook/presto/lark/sheets/SheetsMetadata.java:289]: Gotcha. Thanks!

File: presto-lark-sheets/src/test/java/com/facebook/presto/lark/sheets/TestSheetsUtil.java
Comment (ChunxuTang) [reply presto-lark-sheets/src/test/java/com/facebook/presto/lark/sheets/TestSheetsUtil.java:1]: Did you add them? Forgot to commit changes?
It doesn't show any changes here.

**Review by ChunxuTang — COMMENTED** [review]
@7c00 Thanks for your contributions! 

As you created this PR some time ago and it is now left a bit behind the master branch, could you merge master into this feature branch to avoid conflicts?

File: presto-docs/src/main/sphinx/connector/larksheets.rst
Comment (beinan) [reply presto-docs/src/main/sphinx/connector/larksheets.rst:64]: To make it more readable, I might use a pattern string rather than a real link.  e.g.  https://{account_id}.feishu.ch/sheet/{token}/sheet={sheet_id}

File: presto-lark-sheets/src/main/java/com/facebook/presto/lark/sheets/SheetsConnector.java
Comment (beinan) [inline presto-lark-sheets/src/main/java/com/facebook/presto/lark/sheets/SheetsConnector.java:35]: To be more consistent, can we use LarkSheetsConnector as the class name?  Similarly, other class names should also use LarkSheets

File: presto-lark-sheets/src/test/resources/testing-app-secret-base64
Comment (beinan) [reply presto-lark-sheets/src/test/resources/testing-app-secret-base64:1]: Yes, let's disable the test, otherwise it would be too flaky I guess

We could use ```@Test(enabled = false)```  in the integration test.

Comment (7c00) [reply presto-lark-sheets/src/test/resources/testing-app-secret-base64:1]: Disabled all the tests in `TestLarkSheetsIntegration`.

File: presto-docs/src/main/sphinx/connector/larksheets.rst
Comment (7c00) [reply presto-docs/src/main/sphinx/connector/larksheets.rst]: Updated to a fake app id as `example_app_id`.

File: presto-lark-sheets/src/test/java/com/facebook/presto/lark/sheets/TestSheetsUtil.java
Comment (7c00) [reply presto-lark-sheets/src/test/java/com/facebook/presto/lark/sheets/TestSheetsUtil.java]: The tests was added to another class by mistake. Have corrected this and moved the test to `TestLarkSheetsUtil`.

**Comment by 7c00** [comment]
@ChunxuTang @beinan Thanks for your comments. I have rebased my branch with master and updated the code as the comments. Could you please review it again? Thanks in advance.

File: presto-lark-sheets/src/main/java/com/facebook/presto/lark/sheets/LarkSheetsMetadata.java
Comment (beinan) [inline presto-lark-sheets/src/main/java/com/facebook/presto/lark/sheets/LarkSheetsMetadata.java:285]: I'm just thinking if we could use the column index when the column name is null.   Your call.

**Review by beinan — COMMENTED** [review]
just left a couple comments for some minor issues,  I will continue review this giant pr in this week.  Many thanks for this create contribution!

File: presto-lark-sheets/src/main/java/com/facebook/presto/lark/sheets/LarkSheetsConnector.java
Comment (7c00) [reply presto-lark-sheets/src/main/java/com/facebook/presto/lark/sheets/LarkSheetsConnector.java]: Fixed.

File: presto-lark-sheets/src/main/java/com/facebook/presto/lark/sheets/LarkSheetsErrorCode.java
Comment (7c00) [reply presto-lark-sheets/src/main/java/com/facebook/presto/lark/sheets/LarkSheetsErrorCode.java]: Maybe a typo. Have re-organized the error codes to threw catagories.

File: presto-lark-sheets/src/main/java/com/facebook/presto/lark/sheets/LarkSheetsErrorCode.java
Comment (7c00) [reply presto-lark-sheets/src/main/java/com/facebook/presto/lark/sheets/LarkSheetsErrorCode.java]: Maybe a typo. Have re-orderred the error codes to three catagories.

File: presto-lark-sheets/src/main/java/com/facebook/presto/lark/sheets/LarkSheetsMetadata.java
Comment (7c00) [reply presto-lark-sheets/src/main/java/com/facebook/presto/lark/sheets/LarkSheetsMetadata.java]: Putting the raw tokens in exception messages might cause data leak. Have masked the token before writing the token.

File: presto-lark-sheets/src/main/java/com/facebook/presto/lark/sheets/LarkSheetsMetadata.java
Comment (7c00) [reply presto-lark-sheets/src/main/java/com/facebook/presto/lark/sheets/LarkSheetsMetadata.java:168]: Default schema is not supported, since we dont know which spreadsheet shoud be associated with the schema.

File: presto-lark-sheets/src/main/java/com/facebook/presto/lark/sheets/LarkSheetsMetadata.java
Comment (7c00) [reply presto-lark-sheets/src/main/java/com/facebook/presto/lark/sheets/LarkSheetsMetadata.java:286]: The column name can be a sequence of any UTF-8 characters. So it might be hard to define the column name format. For example, using `$2` as the column index to reference the second column, it might conflict with another column with the name of `$2`. In the other hand, updating a spreadsheet to fill a column name is very easy for end users. So column index name is not be supported.

Comment (beinan) [reply presto-lark-sheets/src/main/java/com/facebook/presto/lark/sheets/LarkSheetsMetadata.java:286]: gotcha, thanks for the clarification!

**Comment by 7c00** [comment]
@ChunxuTang @beinan Hi, thanks for your review. :heart:  Do we need more review before merge this pr?

**Review by beinan — COMMENTED** [review]
I think it looks good to me except two very minor issues.

File: presto-lark-sheets/src/main/java/com/facebook/presto/lark/sheets/LarkSheetsRecordCursor.java
Comment (7c00) [reply presto-lark-sheets/src/main/java/com/facebook/presto/lark/sheets/LarkSheetsRecordCursor.java]: Fixed.

File: presto-lark-sheets/src/test/java/com/facebook/presto/lark/sheets/TestLarkSheetsIntegration.java
Comment (7c00) [reply presto-lark-sheets/src/test/java/com/facebook/presto/lark/sheets/TestLarkSheetsIntegration.java]: This url points to the spreadsheet used for this integration test. The `TESTING_TOKEN` is extracted from the url. I have added comments to make it more clear.

**Comment by 7c00** [comment]
ping @beinan @ChunxuTang

**Review by ChunxuTang — APPROVED** [review]
LGTM. @7c00 Thanks for your nice work!

**Comment by 7c00** [comment]
Rebased with master branch.

**Review by beinan — APPROVED** [review]
lgtm, rerun the test

**Comment by 7c00** [comment]
Thanks @beinan ❤️ . This CI finished. Could we merge this PR?

**Comment by 7c00** [comment]
Thanks @beinan @ChunxuTang for your time to review!

**Comment by rohanpednekar** [comment]
@7c00 Do you think you can provide short documentation on how to use this connector that we can publish at  https://prestodb.io/docs/current/connector.html ?

**Comment by 7c00** [comment]
@rohanpednekar Yes. The connector doc is included in the PR. It's https://github.com/prestodb/presto/blob/master/presto-docs/src/main/sphinx/connector/larksheets.rst. It's a bit strange that it's not displayed in prestodb website.

**Comment by rohanpednekar** [comment]
Thanks @7c00 for the confirmation, I guess this ll get published with the 0.272 release notes PR

**Pull request merged**
